### ISSUE TITLE
Unable to Compile Python Bindings (Linux)

### ISSUE THREAD

**Issue by phoneticsushi** [mgba-emu/mgba#2057]
I'm having multiple issues compiling the Python Bindings:

**My setup:**
* Ubuntu Bionic (18.04) running on Windows 10 WSL (tried WSL 1 and 2 FWIW though don't think that's related)
* Change Line 68 of CmakeLists.txt to: `set(BUILD_PYTHON ON CACHE BOOL "Build Python bindings")` instead of `BUILD_PYTHON OFF`
* Installed necessary packages as described in the Readme

**First Attempt:**
* Building using Docker: `docker run --rm -t -v $PWD:/home/mgba/src mgba/ubuntu:bionic`
* "Python Bindings" presents as "On" in Cmake output but Docker environment appears to be missing a setup for Python:
> CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
PYTHON_LIBRARY (ADVANCED)
    linked by target "mgba" in directory /home/mgba/src

**Second Attempt:**
* Building locally: `cmake -DCMAKE_INSTALL_PREFIX:PATH=/usr ..` from mgba/build dir
* `mgba` target builds successfully, but `mgba-py` target does not:
> [ 96%] Generating lib.c
Traceback (most recent call last):
  File "/home/rweisman/repos/mgba/src/platform/python/_builder.py", line 1, in <module>
    import cffi
ModuleNotFoundError: No module named 'cffi'
python/CMakeFiles/mgba-pylib.dir/build.make:70: recipe for target 'python/lib.c' failed
make[3]: *** [python/lib.c] Error 1
CMakeFiles/Makefile2:376: recipe for target 'python/CMakeFiles/mgba-pylib.dir/all' failed
make[2]: *** [python/CMakeFiles/mgba-pylib.dir/all] Error 2
CMakeFiles/Makefile2:426: recipe for target 'python/CMakeFiles/mgba-py.dir/rule' failed
make[1]: *** [python/CMakeFiles/mgba-py.dir/rule] Error 2
Makefile:249: recipe for target 'mgba-py' failed
make: *** [mgba-py] Error 2
* Not sure why this happens as I have `cffi` installed (via `python3 -m pip install cffi`) and I can `import cffi` just fine elsewhere

**Third Attempt:**
* Run `_builder.py` manually from `mgba/src/platform/python`
* No `cffi` issues this way: the output is: `generating lib.c` and it does generate said file
* Replace `_builder.py` with an empty file to work around the `import cffi` issue
* Return to `mgba/build` and run `make` again
* This builds `mgba-pylib` correctly but `mgba-py` gets stuck since it depends on the original `_builder.py`
* Replace original `_builder.py` file
* Linker errors building `mgba-sdl`:
> [ 58%] Linking C executable mgba
../python/libmgba-pylib.a(engine.c.o): In function `mPythonScriptDebuggerEntered':
engine.c:(.text+0x3d): undefined reference to `mPythonSetDebugger'
../python/libmgba-pylib.a(engine.c.o): In function `mPythonScriptEngineRun':
engine.c:(.text+0x86): undefined reference to `mPythonSetDebugger'
../python/libmgba-pylib.a(engine.c.o): In function `mPythonScriptDebuggerEntered':
engine.c:(.text+0x4d): undefined reference to `mPythonDebuggerEntered'
../python/libmgba-pylib.a(engine.c.o): In function `mPythonScriptEngineLookupSymbol':
engine.c:(.text+0x67): undefined reference to `mPythonLookupSymbol'
../python/libmgba-pylib.a(engine.c.o): In function `mPythonScriptEngineRun':
engine.c:(.text+0x91): undefined reference to `mPythonRunPending'
../python/libmgba-pylib.a(engine.c.o): In function `mPythonScriptEngineLoadScript':
engine.c:(.text+0xa7): undefined reference to `mPythonLoadScript'
collect2: error: ld returned 1 exit status
sdl/CMakeFiles/mgba-sdl.dir/build.make:263: recipe for target 'sdl/mgba' failed
make[2]: *** [sdl/mgba] Error 1
CMakeFiles/Makefile2:566: recipe for target 'sdl/CMakeFiles/mgba-sdl.dir/all' failed
make[1]: *** [sdl/CMakeFiles/mgba-sdl.dir/all] Error 2
Makefile:162: recipe for target 'all' failed
make: *** [all] Error 2

**tl;dr:**
* Docker environment not set up to compile Python bindings
* Target `mgba-pylib` can't see `cffi` for some reason even though it's installed
* Linker errors when building target `mgba-sdl`

I've tried this entire process on both of the following commits with the same effect:
* `7d672a2215643d205e0edad39e1903e49edd70d0` (`master` as of writing)
* `339c0145c94e9ca03dec6e8987c280563b9515e8` (tag `0.8` as of writing)

Is there some setting/documentation I've missed?  Happy to try alternate configurations/post full output if that would be helpful in troubleshooting.

**Comment by endrift** [comment]
That's...not how you're supposed to change those settings, but that shouldn't matter here.

That said, it looks like you're definitely missing _something_ but I have no idea what. I've never seen problems like this. Your second attempt looks closer to what you're supposed to do than the first, but still...

**Comment by endrift** [comment]
Following these steps inside a docker container based on the bionic image seems to work (when started with --user root):

```
# apt update && apt install libpython3-dev python3-venv
# su - mgba
$ bash
$ mkdir -p ~/src/build-bionic
$ cd ~/src/build-bionic
$ python3 -m venv venv
$ . ./venv/bin/activate
$ pip install cffi pytest-runner cached-property
$ cmake .. -DBUILD_PYTHON=ON
$ make
```

That's not really documented but the bindings aren't particularly supported at the moment and the virtualenv probably isn't needed, but this was what I tried.

**Comment by rweisman** [comment]
First, thank you for getting back to me so quickly.  Second, could you tell me the recommended way to change those settings?  Figured there was a proper way to do so but didn't see it in the documentation - that may resolve the entire issue.

In the meantime, I'll try the steps you posted above within the Docker container on my end and report back shortly.

**Comment by endrift** [comment]
The proper way is what I did here: -D[setting name]=[value]. That's just how to use CMake from the command line. You can't do that directly to the Docker command (you need to pass it through an environment variable) but given the image doesn't have Python installed anyway that doesn't hugely affect that command anyway.

**Comment by rweisman** [comment]
Apologies for the delay - first time using Docker.

That builds all targets without error, including `mgba-pylib` and `mgba-py`, but doesn't generate a `setup.py` file as described [here](https://github.com/mgba-emu/mgba/issues/997#issuecomment-368132640)
> (venv) mgba@353894b8533f: ~/src/build-bionic$ ls python/
CMakeFiles  CTestTestfile.cmake  Makefile  cmake_install.cmake  lib.c  lib.linux-x86_64-3.6  libmgba-pylib.a  temp.linux-x86_64-3.6
(venv) mgba@353894b8533f: ~/src/build-bionic$ find . -name "setup.py"
(venv) mgba@353894b8533f: ~/src/build-bionic$

**Comment by endrift** [comment]
Yes, it's a bit complicated, unfortunately. If you want to install it into the virtualenv you should use `make mgba-py-install`, though that _does_ seem to be broken right now. I suppose I'm curious what your use case for the bindings is, because it might not currently be supported.

**Comment by rweisman** [comment]
The goal is to be able to replace the soundtrack of a Game Boy (Color) game while it's being played.  Put another way, a playable version of [what's shown in this video](https://www.youtube.com/watch?v=mdiXsPTZ3QM) (incidentally, the inspiration for the project - would highly recommend this album in particular; the composer's orchestration is amazing).

To accomplish that goal, would need to be able to do the following:
* Play a game normally (input buttons, display screen, play audio)
* Read/write emulated registers
* Read/write emulated RAM
* Read active emulated RAM Bank value

Setting breakpoints and receiving callbacks on them would be ideal, but I'm not above polling the emulated PC on every step so long as I can read the registers.

Planning to publish a playable toy example to GitHub when it works.  Might eventually become a framework for replacing game music in arbitrary games/platforms somewhere in the far future... but first things first.

**Comment by endrift** [comment]
These bindings cannot interact with the existing Qt frontend ad this point, so I'm not sure you'll be able to do what you want.

**Comment by rweisman** [comment]
If you say it can't be done (yet), I believe you 100%.  Confused by that description though - seeing the functions `raw_read` and `raw_write` in the MemoryView as well as `set_breakpoint` and `add_callback` in the NativeDebugger, I was under the impression that the bindings would interact directly with a Core, which would be controlled/rendered separately in other threads by Qt/SDL (e.g.).

Is this sort of thing possible/intended via C/C++, and the Python bindings are WIP, or do I just completely misunderstand the application's architecture?

**Comment by endrift** [comment]
The problem is that the bindings are mostly useful for running the core directly, not as part of the existing frontends. You can load scripts in the SDL frontend (maybe, I'm not sure how well that works), but you absolutely cannot do it in the Qt frontend yet. The bindings are very WIP and are likely to change dramatically in the future.

**Comment by rweisman** [comment]
Quite understandable and thanks for answering all my questions.  Please let me know if you'd like any help on that front - can't promise anything of course but would love to contribute if possible.

Feel free to close this ticket or leave it open as a reminder to look at the `mgba-py-install` target at some point.

I hope this thread is useful to anyone else who encounters the project looking for Python bindings.

**Comment by RetroEdit** [comment]
Since the current Python bindings are now deprecated in favor of the newer Lua scripting subsystem, I'm going to close this.

Future progress on Python bindings can be tracked in #3048.
